### PR TITLE
[Temporal] Add link prediction example

### PR DESCRIPTION
**Pull request by peizhou001**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/sampling/graphbolt/temporal_link_prediction.py [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+    def forward(self, input_feat: torch.Tensor):
+        return self.embed(input_feat.view(-1))
+
+
+class HeteroSAGE(nn.Module):
+    def __init__(self, in_size, hidden_size):
+        super().__init__()
+        self.layers = nn.ModuleList()
+        self.layers.append(
+            dglnn.HeteroGraphConv(
```
Comment (mfbalin) [inline examples/sampling/graphbolt/temporal_link_prediction.py:84]: Can we add these two layers in a loop? It looks like only the dimensions are changing.

File: examples/sampling/graphbolt/temporal_link_prediction.py [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+        features, node_feature_keys=node_feature_keys
+    )
+
+    ############################################################################
+    # [Input]:
+    # 'device': The device to copy the data to.
+    # [Output]:
+    # A CopyTo object to copy the data to the specified device.
+    ############################################################################
+    if args.storage_device == "cpu":
```
Comment (mfbalin) [inline examples/sampling/graphbolt/temporal_link_prediction.py:198]: Moving copy_to before fetch feature will improve performance a lot.

We need to pin the features when there is a gpu though.

File: examples/sampling/graphbolt/temporal_link_prediction.py [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+    args.device = torch.device(args.device)
+
+    # Load and preprocess dataset.
+    print("Loading data")
+    dataset_path = download_datasets(args.dataset)
+    dataset = gb.OnDiskDataset(dataset_path).load()
+
+    # Move the dataset to the selected storage.
+    graph = dataset.graph.to(args.storage_device)
+    features = dataset.feature.to(args.storage_device)
```
Comment (mfbalin) [inline examples/sampling/graphbolt/temporal_link_prediction.py:346]: Let's pin the features when there is a gpu. Together with moving copy_to before fetch feature, can improve performance a lot.

Comment (mfbalin) [reply examples/sampling/graphbolt/temporal_link_prediction.py:346]: You can check the PyG advanced example for the different placement of graph, features etc.

**Review by mfbalin — APPROVED** [review]
LGTM overall.

File: examples/sampling/graphbolt/temporal_link_prediction.py [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+prediction with temporal information using graphbolt dataloader.
+
+Unlike previous dgl examples, we've utilized the newly defined dataloader
+from GraphBolt. This example will help you grasp how to build an end-to-end
+training pipeline using GraphBolt.
+
+While node classification predicts labels for nodes based on their
+local neighborhoods, link prediction assesses the likelihood of an edge
+existing between two nodes, necessitating different sampling strategies
+that account for pairs of nodes and their joint neighborhoods.
```
Comment (frozenbugs) [inline examples/sampling/graphbolt/temporal_link_prediction.py:12]: Unrelated, can be removed, maybe talk more about how what's temporal in the comment.

File: examples/sampling/graphbolt/temporal_link_prediction.py (hunk 2 above)
Comment (peizhou001) [reply examples/sampling/graphbolt/temporal_link_prediction.py:198]: Currently an segment error will raise if use pinned memory, I will add this after the bug fix.

File: examples/sampling/graphbolt/temporal_link_prediction.py (hunk 3 above)
Comment (peizhou001) [reply examples/sampling/graphbolt/temporal_link_prediction.py:346]: See above.

File: examples/sampling/graphbolt/temporal_link_prediction.py [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+from dgl.data.utils import download, extract_archive
+
+
+TIMESTAMP_FEATURE_NAME = "__timestamp__"
+node_feature_keys = {
+    "Product": ["categoryId"],
+    "Query": ["categoryId"],
+}
+
+target_type = ("Query", "Click", "Product")
```
Comment (frozenbugs) [inline examples/sampling/graphbolt/temporal_link_prediction.py:53]: TARGET_ETYPE / SEED_ETYPE?

File: examples/sampling/graphbolt/temporal_link_prediction.py [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+            is_last_layer = layer_idx == len(self.layers) - 1
+            if not is_last_layer:
+                H_node_dict = {
+                    ntype: F.relu(H) for ntype, H in H_node_dict.items()
+                }
+        return H_node_dict
+
+
+def create_dataloader(args, graph, features, itemset, is_train=True):
+    """Get a GraphBolt version of a dataloader for link prediction tasks. This
```
Comment (frozenbugs) [inline examples/sampling/graphbolt/temporal_link_prediction.py:120]: Remove graphbolt related comments, this example is not for teaching people using GB, and should focus on Temporal related code.

File: examples/sampling/graphbolt/temporal_link_prediction.py [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+        "--mode",
+        default="cpu-cuda",
+        choices=["cpu-cpu", "cpu-cuda"],
+        help="Dataset storage placement and Train device: 'cpu' for CPU and RAM,"
+        " 'pinned' for pinned memory in RAM, 'cuda' for GPU and GPU memory.",
+    )
+    return parser.parse_args()
+
+
+def download_datasets(name, root="datasets"):
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply examples/sampling/graphbolt/temporal_link_prediction.py:308]: Discussed with @Rhett-Ying , the dataset is not well confirmed yet, will move it to Built-in later.

File: examples/sampling/graphbolt/temporal_link_prediction.py [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+
+    train_set = dataset.tasks[0].train_set
+    args.fanout = list(map(int, args.fanout.split(",")))
+
+    in_size = 128
+    hidden_channels = 256
+    args.device = torch.device(args.device)
+    model = HeteroSAGE(in_size, hidden_channels).to(args.device)
+    encoders = {
+        "Query": CategoricalEncoder(1102, in_size).to(args.device),
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply examples/sampling/graphbolt/temporal_link_prediction.py:345]: The dim for input feature size, separately `Query:category` and `Product:category`

File: examples/sampling/graphbolt/temporal_link_prediction.py (hunk 2 above)
Comment (mfbalin) [reply examples/sampling/graphbolt/temporal_link_prediction.py:187]: Can you paste the error?

File: examples/sampling/graphbolt/temporal_link_prediction.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (frozenbugs) [reply examples/sampling/graphbolt/temporal_link_prediction.py:308]: OK, add a TODO here.

File: examples/sampling/graphbolt/temporal_link_prediction.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (frozenbugs) [reply examples/sampling/graphbolt/temporal_link_prediction.py:345]: Better to refer from the Tensor.size instead of hard code.

File: examples/sampling/graphbolt/temporal_link_prediction.py (hunk 5 above)
Comment (peizhou001) [reply examples/sampling/graphbolt/temporal_link_prediction.py:53]: TARGET_ETYPE, not only about seeds, also about target mfg.

File: examples/sampling/graphbolt/temporal_link_prediction.py [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+                        hidden_size,
+                        hidden_size,
+                        "mean",
+                    )
+                    for etype in all_connected_types
+                },
+                aggregate="sum",
+            )
+        )
+        self.hidden_size = hidden_size
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply examples/sampling/graphbolt/temporal_link_prediction.py:98]: removed

File: examples/sampling/graphbolt/temporal_link_prediction.py [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+                ].squeeze()
+
+            blocks = data.blocks
+
+            # Get the embeddings of the input nodes.
+            H_node_dict = {
+                ntype: encoders[ntype](feat)
+                for ntype, feat in node_feature.items()
+            }
+            seed_embeds = model(blocks, H_node_dict)
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply examples/sampling/graphbolt/temporal_link_prediction.py:237]: use `X_node_dict`, as `X` denotes feature/embedding and `dict ` implictly shows it contains multiple type of nodes.

File: examples/sampling/graphbolt/temporal_link_prediction.py (hunk 2 above)
Comment (peizhou001) [reply examples/sampling/graphbolt/temporal_link_prediction.py:181]: 图片

File: examples/sampling/graphbolt/temporal_link_prediction.py [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+            is_last_layer = layer_idx == len(self.layers) - 1
+            if not is_last_layer:
+                H_node_dict = {
+                    ntype: F.relu(H) for ntype, H in H_node_dict.items()
+                }
+        return H_node_dict
+
+
+def create_dataloader(args, graph, features, itemset, is_train=True):
+    ############################################################################
```
Comment (frozenbugs) [inline examples/sampling/graphbolt/temporal_link_prediction.py:119]: Remove all this ##### comment if not related to temporal

**Review by frozenbugs — APPROVED** [review]
